File: app/src/main/java/com/moneywise/data/Calculators.kt
package com.moneywise.data

import java.text.NumberFormat
import java.util.Locale

data class WorkTimeResult(
    val hours: Double,
    val minutes: Int,
    val grossHours: Double,
    val grossMinutes: Int,
    val humanReadable: String,
    val humanReadableGross: String
)

data class SavingsYear(
    val year: Int,
    val totalBalance: Double,
    val contributed: Double,
    val interestEarned: Double
)

data class SavingsResult(
    val initialAmount: Double,
    val monthlyContribution: Double,
    val annualRate: Double,
    val years: Int,
    val timeline: List<SavingsYear>
) {
    val finalAmount: Double get() = timeline.lastOrNull()?.totalBalance ?: initialAmount
    val totalContributed: Double get() = initialAmount + (monthlyContribution * years * 12)
    val totalInterest: Double get() = finalAmount - totalContributed
}

data class InvestmentYear(
    val year: Int,
    val nominalBalance: Double,
    val realBalance: Double,
    val contributed: Double
)

data class InvestmentResult(
    val initialAmount: Double,
    val monthlyContribution: Double,
    val annualReturn: Double,
    val years: Int,
    val inflationRate: Double,
    val timeline: List<InvestmentYear>
) {
    val finalAmount: Double get() = timeline.lastOrNull()?.nominalBalance ?: initialAmount
    val finalRealAmount: Double get() = timeline.lastOrNull()?.realBalance ?: initialAmount
    val totalContributed: Double get() = initialAmount + (monthlyContribution * years * 12)
    val totalGain: Double get() = finalAmount - totalContributed
    val totalRealGain: Double get() = finalRealAmount - totalContributed
}

object Calculators {

    fun calculateWorkTime(purchasePrice: Double, profile: SalaryProfile): WorkTimeResult {
        val grossHours = purchasePrice / profile.hourlyWage
        val netHours = purchasePrice / profile.netHourly

        val grossH = grossHours.toInt()
        val grossM = ((grossHours - grossH) * 60).toInt()
        val netH = netHours.toInt()
        val netM = ((netHours - netH) * 60).toInt()

        val hoursPerDay = profile.hoursPerWeek / 5

        val humanReadable = when {
            netHours < 1 -> "${netM} minuten"
            netHours < 24 -> "$netH uur en $netM minuten"
            else -> "${(netHours / hoursPerDay).toString().take(4)} werkdagen"
        }

        val humanReadableGross = when {
            grossHours < 1 -> "${grossM} minuten"
            grossHours < 24 -> "$grossH uur en $grossM minuten"
            else -> "${(grossHours / hoursPerDay).toString().take(4)} werkdagen"
        }

        return WorkTimeResult(
            hours = netHours,
            minutes = netM,
            grossHours = grossHours,
            grossMinutes = grossM,
            humanReadable = humanReadable,
            humanReadableGross = humanReadableGross
        )
    }

    fun calculateSavings(
        initialAmount: Double,
        monthlyContribution: Double,
        annualRate: Double,
        years: Int
    ): SavingsResult {
        val monthlyRate = annualRate / 100 / 12
        val timeline = mutableListOf<SavingsYear>()
        var balance = initialAmount
        var totalContributed = initialAmount

        for (year in 1..years) {
            repeat(12) {
                balance += monthlyContribution
                balance *= (1 + monthlyRate)
                totalContributed += monthlyContribution
            }
            timeline.add(SavingsYear(
                year = year,
                totalBalance = balance,
                contributed = totalContributed,
                interestEarned = balance - totalContributed
            ))
        }

        return SavingsResult(
            initialAmount = initialAmount,
            monthlyContribution = monthlyContribution,
            annualRate = annualRate,
            years = years,
            timeline = timeline
        )
    }

    fun calculateInvestment(
        initialAmount: Double,
        monthlyContribution: Double,
        annualReturn: Double,
        years: Int,
        inflationRate: Double = 2.5
    ): InvestmentResult {
        val monthlyReturn = annualReturn / 100 / 12
        val monthlyInflation = inflationRate / 100 / 12
        val timeline = mutableListOf<InvestmentYear>()
        var nominalBalance = initialAmount
        var realBalance = initialAmount
        var totalContributed = initialAmount

        for (year in 1..years) {
            repeat(12) {
                nominalBalance += monthlyContribution
                nominalBalance *= (1 + monthlyReturn)
                realBalance += monthlyContribution
                realBalance *= (1 + monthlyReturn - monthlyInflation)
                totalContributed += monthlyContribution
            }
            timeline.add(InvestmentYear(
                year = year,
                nominalBalance = nominalBalance,
                realBalance = realBalance,
                contributed = totalContributed
            ))
        }

        return InvestmentResult(
            initialAmount = initialAmount,
            monthlyContribution = monthlyContribution,
            annualReturn = annualReturn,
            years = years,
            inflationRate = inflationRate,
            timeline = timeline
        )
    }

    fun formatCurrency(amount: Double, currency: Currency, showDecimals: Boolean = true): String {
        val format = NumberFormat.getCurrencyInstance(Locale.getDefault())
        format.currency = java.util.Currency.getInstance(currency.code)
        if (!showDecimals) format.maximumFractionDigits = 0
        return format.format(amount)
    }

    fun formatCompact(amount: Double, currency: Currency): String {
        return when {
            amount >= 1_000_000 -> "${currency.symbol}${String.format("%.1f", amount / 1_000_000)}M"
            amount >= 1_000 -> "${currency.symbol}${String.format("%.1f", amount / 1_000)}K"
            else -> formatCurrency(amount, currency)
        }
    }
}

File: lib/utils/calculators.dart
import '../models/salary_profile.dart';

class WorkTimeResult {
  final double hours;
  final double minutes;
  final double grossHours;
  final double grossMinutes;
  final double costGross;
  final double costNet;
  final String humanReadable;
  final String humanReadableGross;

  const WorkTimeResult({
    required this.hours,
    required this.minutes,
    required this.grossHours,
    required this.grossMinutes,
    required this.costGross,
    required this.costNet,
    required this.humanReadable,
    required this.humanReadableGross,
  });
}

class SavingsResult {
  final double initialAmount;
  final double monthlyContribution;
  final double annualRate;
  final int years;
  final List<SavingsYear> timeline;

  const SavingsResult({
    required this.initialAmount,
    required this.monthlyContribution,
    required this.annualRate,
    required this.years,
    required this.timeline,
  });

  double get finalAmount => timeline.isNotEmpty ? timeline.last.totalBalance : initialAmount;
  double get totalContributed => initialAmount + (monthlyContribution * years * 12);
  double get totalInterest => finalAmount - totalContributed;
}

class SavingsYear {
  final int year;
  final double totalBalance;
  final double contributed;
  final double interestEarned;

  const SavingsYear({
    required this.year,
    required this.totalBalance,
    required this.contributed,
    required this.interestEarned,
  });
}

class InvestmentResult {
  final double initialAmount;
  final double monthlyContribution;
  final double annualReturn;
  final int years;
  final double inflationRate;
  final List<InvestmentYear> timeline;

  const InvestmentResult({
    required this.initialAmount,
    required this.monthlyContribution,
    required this.annualReturn,
    required this.years,
    required this.inflationRate,
    required this.timeline,
  });

  double get finalAmount => timeline.isNotEmpty ? timeline.last.nominalBalance : initialAmount;
  double get finalRealAmount => timeline.isNotEmpty ? timeline.last.realBalance : initialAmount;
  double get totalContributed => initialAmount + (monthlyContribution * years * 12);
  double get totalGain => finalAmount - totalContributed;
  double get totalRealGain => finalRealAmount - totalContributed;
}

class InvestmentYear {
  final int year;
  final double nominalBalance;
  final double realBalance;
  final double contributed;

  const InvestmentYear({
    required this.year,
    required this.nominalBalance,
    required this.realBalance,
    required this.contributed,
  });
}

class Calculators {
  static WorkTimeResult calculateWorkTime({
    required double purchasePrice,
    required SalaryProfile profile,
  }) {
    final grossHours = purchasePrice / profile.hourlyWage;
    final netHours = purchasePrice / profile.netHourly;

    final grossH = grossHours.floor();
    final grossM = ((grossHours - grossH) * 60).round();
    final netH = netHours.floor();
    final netM = ((netHours - netH) * 60).round();

    String humanReadable;
    if (netHours < 1) {
      humanReadable = '${netM.round()} minuten';
    } else if (netHours < 24) {
      humanReadable = '$netH uur en $netM minuten';
    } else {
      final days = (netHours / (profile.hoursPerWeek / 5)).toStringAsFixed(1);
      humanReadable = '$days werkdagen';
    }

    String humanReadableGross;
    if (grossHours < 1) {
      humanReadableGross = '${grossM.round()} minuten';
    } else if (grossHours < 24) {
      humanReadableGross = '$grossH uur en $grossM minuten';
    } else {
      final days = (grossHours / (profile.hoursPerWeek / 5)).toStringAsFixed(1);
      humanReadableGross = '$days werkdagen';
    }

    return WorkTimeResult(
      hours: netHours,
      minutes: netM.toDouble(),
      grossHours: grossHours,
      grossMinutes: grossM.toDouble(),
      costGross: purchasePrice / profile.hourlyWage * profile.hourlyWage,
      costNet: purchasePrice,
      humanReadable: humanReadable,
      humanReadableGross: humanReadableGross,
    );
  }

  static SavingsResult calculateSavings({
    required double initialAmount,
    required double monthlyContribution,
    required double annualRate,
    required int years,
  }) {
    final monthlyRate = annualRate / 100 / 12;
    final timeline = <SavingsYear>[];
    double balance = initialAmount;
    double totalContributed = initialAmount;

    for (int year = 1; year <= years; year++) {
      for (int month = 0; month < 12; month++) {
        balance += monthlyContribution;
        balance *= (1 + monthlyRate);
        totalContributed += monthlyContribution;
      }
      timeline.add(SavingsYear(
        year: year,
        totalBalance: balance,
        contributed: totalContributed,
        interestEarned: balance - totalContributed,
      ));
    }

    return SavingsResult(
      initialAmount: initialAmount,
      monthlyContribution: monthlyContribution,
      annualRate: annualRate,
      years: years,
      timeline: timeline,
    );
  }

  static InvestmentResult calculateInvestment({
    required double initialAmount,
    required double monthlyContribution,
    required double annualReturn,
    required int years,
    double inflationRate = 2.5,
  }) {
    final monthlyReturn = annualReturn / 100 / 12;
    final monthlyInflation = inflationRate / 100 / 12;
    final timeline = <InvestmentYear>[];
    double nominalBalance = initialAmount;
    double realBalance = initialAmount;
    double totalContributed = initialAmount;

    for (int year = 1; year <= years; year++) {
      for (int month = 0; month < 12; month++) {
        nominalBalance += monthlyContribution;
        nominalBalance *= (1 + monthlyReturn);
        realBalance += monthlyContribution;
        realBalance *= (1 + monthlyReturn - monthlyInflation);
        totalContributed += monthlyContribution;
      }
      timeline.add(InvestmentYear(
        year: year,
        nominalBalance: nominalBalance,
        realBalance: realBalance,
        contributed: totalContributed,
      ));
    }

    return InvestmentResult(
      initialAmount: initialAmount,
      monthlyContribution: monthlyContribution,
      annualReturn: annualReturn,
      years: years,
      inflationRate: inflationRate,
      timeline: timeline,
    );
  }

  static String formatWorkTime(double totalHours) {
    final h = totalHours.floor();
    final m = ((totalHours - h) * 60).round();
    if (h == 0) return '$m min';
    if (m == 0) return '$h uur';
    return '$h uur en $m min';
  }
}
